server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -540,25 +555,33 @@ private static Map<String, Mapper> buildMergedMappers(
             ObjectMapper existing,
             ObjectMapper mergeWithObject,
             MergeReason reason,
-            MapperMergeContext objectMergeContext
+            MapperMergeContext objectMergeContext,
+            boolean subobjects
         ) {
-            Iterator<Mapper> iterator = mergeWithObject.iterator();
-            if (iterator.hasNext() == false) {
-                return Map.copyOf(existing.mappers);
+            Map<String, Mapper> mergedMappers = new HashMap<>();
+            for (Mapper existingMapper : existing.mappers.values()) {
+                if (subobjects == false && existingMapper instanceof ObjectMapper objectMapper) {
+                    objectMapper.flatten(objectMergeContext.getMapperBuilderContext()).forEach(m -> mergedMappers.put(m.simpleName(), m));
+                } else {
+                    putMergedMapper(mergedMappers, existingMapper);
+                }
```

Comment:
I am trying to summarize in my head why we need to auto-flatten at merge time, and not only when getting mappings. What would not work if we did not do it at merge time perhaps?

Reply:
Perhaps the answer is #103497 .

Reply:
It's also when two independent component templates are combined in the context of an index template. The component templates may not be written with subobjects: false in mind. But when they're imported into an index template that sets subobjects: false, the component templates need to be flattened during that merge process.